cmd/perf/internal/managed/managed.go

```diff
@@ -0,0 +1,307 @@
+// Copyright 2022 The Crossplane Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managed
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"gopkg.in/yaml.v2"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"k8s.io/client-go/dynamic"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/upbound/uptest/cmd/perf/internal/common"
+)
+
+// RunExperiment runs the experiment according to command-line inputs.
+// Firstly the input manifests are deployed. After the all MRs are ready, time to readiness metrics are calculated.
+// Then, by default, all deployed MRs are deleted.
+func RunExperiment(mrTemplatePaths map[string]int, clean bool) ([]common.Result, error) { //nolint:gocyclo
+	var timeToReadinessResults []common.Result //nolint:prealloc
+
+	client := createDynamicClient()
+
+	if err := applyResources(client, mrTemplatePaths); err != nil {
+		return nil, err
+	}
+
+	if err := checkReadiness(client, mrTemplatePaths); err != nil {
+		return nil, err
+	}
+
+	timeToReadinessResults, err := calculateReadinessDuration(client, mrTemplatePaths)
+	if err != nil {
+		return nil, err
+	}
+
+	if clean {
+		log.Info("Deleting resources...")
+		if err := deleteResources(client, mrTemplatePaths); err != nil {
+			return nil, err
+		}
+		log.Info("Checking deletion of resources...")
+		if err := checkDeletion(client, mrTemplatePaths); err != nil {
+			return nil, err
+		}
+	}
+	return timeToReadinessResults, nil
+}
+
+func applyResources(client dynamic.Interface, mrTemplatePaths map[string]int) error {
+	file, err := os.Create(fmt.Sprintf("/tmp/test.yaml"))
+	if err != nil {
+		return err
+	}
+	if err := file.Close(); err != nil {
+		return err
+	}
+
+	for mrPath, count := range mrTemplatePaths {
+		m, err := readYamlFile(mrPath)
+		if err != nil {
+			return err
+		}
+		o := prepareUnstructuredObject(m)
+
+		f, err := os.OpenFile("/tmp/test.yaml", os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+		if err != nil {
+			return err
+		}
+
+		for i := 1; i <= count; i++ {
+			o["metadata"].(map[string]interface{})["name"] = fmt.Sprintf("test-%d", i)
```

Comment:
There are some resources (for example, Azure Registry) that do not allow for hyphen in the name. Maybe making the function more flexible or just dropping the hyphen from the name could help?